Documentation/benchmarks/etcd-3-watch-memory-benchmark.md

```diff
@@ -0,0 +1,72 @@
+# Watch Memory Usage Benchmark
+
+*NOTE*: The watch features are still being actively developed and we expect the memory usage to change over time. However we do expect that memory usage should not be significantly increased from the figures below.
+
+A primary goal of etcd is to support large numbers of watching. etcd aims to support as many as O(10k) clients, O(100K) watch streams (O(10) streams per client) and O(10M) total watchings (O(100) watching per stream). Thus the goal is to minimize the memory consumption per watching, since it will dominate the overall consumption.
+
+Three related components of etcd watch consume physical memory: grpc.Conn, watch stream, watching. grpc.Conn maintains the actual TCP connection and other gRPC connection state. Each grpc.Conn consumes O(10kb) of memory.  Each grpc.Conn might have multiple watch streams attached. 
+Each watch stream is an independent HTTP2 connection that consumes O(10kb) memory. 
+Multiple watchings might also share one watch stream. Each watching should consume < O(1kb).
+
+```
+                                          +-------+
+                                          | watch |
+                              +---------> | foo   |
+                              |           +-------+
+                       +------+-----+
+                       |   stream   |
+      +--------------> |            |
+      |                +------+-----+     +-------+
+      |                       |           | watch |
+      |                       +---------> | bar   |
++-----+------+                            +-------+
+|            |         +------------+
+|   conn     +-------> |   stream   |
+|            |         |            |
++-----+------+         +------------+
+      |
+      |
+      |
+      |                +------------+
+      +--------------> |   stream   |
+                       |            |
+                       +------------+
+```
+
+The theoretical memory consumption of watch can be approximated with the formula:
+`memory = c1 * number_of_conn + c2 * avg_number_of_stream_per_conn + c3 * avg_number_of_watch_stream`
+
+## Testing Environment
+
+etcd version
+- git head https://github.com/coreos/etcd/commit/185097ffaa627b909007e772c175e8fefac17af3
+
+GCE n1-standard-2 machine type
+- 7.5 GB memory
+- 2x CPUs
+
+## Overall memory usage
+
+The overall memory usage captures how much RSS etcd consumes with the client watches. The result is not very accurate and might vary around 10%. It is still very meaningful, since the goal is to learn about the rough memory usage and the pattern (find out `c1`, `c2` and `c3`).
```

Comment:
Link to wiki or something?

On Fri, Nov 13, 2015 at 2:44 PM, Xiang Li <notifications@github.com> wrote:

> In Documentation/benchmarks/etcd-3-watch-memory-benchmark.md
> <https://github.com/coreos/etcd/pull/3862#discussion_r44843152>:
>
> > +
> > +The theoretical memory consumption of watch can be approximated with the formula:
> > +`memory = c1 * number_of_conn + c2 * avg_number_of_stream_per_conn + c3 * avg_number_of_watch_stream`
> > +
> > +## Testing Environment
> > +
> > +etcd version
> > +- git head https://github.com/coreos/etcd/commit/185097ffaa627b909007e772c175e8fefac17af3
> > +
> > +GCE n1-standard-2 machine type
> > +- 7.5 GB memory
> > +- 2x CPUs
> > +
> > +## Overall memory usage
> > +
> > +The overall memory usage captures how much RSS etcd consumes with the client watches. The result is not very accurate and might vary around 10%. It is still very meaningful, since the goal is to learn about the rough memory usage and the pattern (find out `c1`, `c2` and `c3`).
>
> probably I should add a note or something?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/3862/files#r44843152>.
>